changes/2020-07-17_cache-entry-identifier-formulas/change.md

```diff
@@ -0,0 +1,177 @@
+
+Some of the cache entry identifier formulas include
+the two-byte algorithm suite ID for the algorithm suite in a materials request.
+The algorithm suite IDs are defined in the
+[Supported Algorithm Suites section of the Algorithm Suites specification](../../framework/algorithm-suites.md#supported-algorithm-suites).
+In the following appendix sections we use `AlgorithmSuiteId`
+to the denote the function that,
+given an algorithm suite as specified in a materials request,
+returns the corresponding two-byte algorithm suite ID.
+
+### Encryption Materials, Without Algorithm Suite
+
+If the Get Encryption Materials request does not specify an algorithm suite,
+then the cache entry identifier MUST be calculated
+as the SHA-512 hash of the concatenation of the following byte strings,
+in the order listed:
+
+1. The SHA-512 hash of the caching CMM’s Partition ID
+2. One null byte (`0x00`)
+3. The SHA-512 hash of the serialized encryption context
+
+As a formula:
+
+```
+ENTRY_ID = SHA512(
+    SHA512(cachingCMM.partitionId)
+    + 0x00
+    + SHA512(SerializeEncryptionContext(getEncryptionMaterialsRequest.encryptionContext))
+)
+```
+
+### Encryption Materials, With Algorithm Suite
+
+If the Get Encryption Materials request does specify an algorithm suite,
+then the cache entry identifier MUST be calculated
+as the SHA-512 hash of the concatenation of the following byte strings,
+in the order listed:
+
+1.  The SHA-512 hash of the caching CMM’s Partition ID
+2.  One byte with value 1 (`0x01`)
+3.  The two-byte algorithm suite ID corresponding to the algorithm suite in the request
+4.  The SHA-512 hash of the serialized encryption context
+
+As a formula:
+
+```
+ENTRY_ID = SHA512(
+    SHA512(cachingCMM.partitionId)
+    + 0x01
+    + AlgorithmSuiteId(getEncryptionMaterialsRequest.algorithmSuite)
+    + SHA512(SerializeEncryptionContext(getEncryptionMaterialsRequest.encryptionContext))
+)
+```
+
+### Decryption Materials
+
+When the caching CMM receives a Decrypt Materials request,
+it MUST calculate the cache entry identifier as
+the SHA-512 hash of the concatenation of the following byte strings,
+
+in the order listed:
+
+1.  The SHA-512 hash of the caching CMM’s Partition ID
+2.  The two-byte algorithm suite ID corresponding to the algorithm suite in the request
+3.  The concatenation of the lexicographically-sorted SHA-512 hashes of the serialized encrypted data keys,
+    where serialization is as defined in the Message Header Encrypted Data Key Entries specification.
+4.  A sentinel field of 512 zero bits (or equivalently, 64 null bytes), indicating the end of the key hashes
+5.  The SHA-512 hash of the serialized encryption context
+
+As a formula:
+
+```
+ENTRY_ID = SHA512(
+    SHA512(cachingCMM.partitionId)
+    + AlgorithmSuiteId(decryptMaterialsRequest.algorithmSuite)
+    + CONCATENATE(SORTED(
+        SHA512(SerializeEncryptedDataKey(encryptedDataKey))
+        for encryptedDataKey in decryptMaterialsRequest.encryptedDataKeys
+    ))
```

Comment:
Is there a way to make this slightly clearer? Maybe break the following up into a separate line above `ENTRY_ID` with named variables?
```
        SHA512(SerializeEncryptedDataKey(encryptedDataKey))
        for encryptedDataKey in decryptMaterialsRequest.encryptedDataKeys
```

Basically, we're looping through every encryptedDataKey and determining `SHA512(SerializeEncryptedDataKey(encryptedDataKey))` for each. We then add them to a sorted list and finally concatenate it. This just wasn't super clear to me at first.

Reply:
Yeah, the formula is a bit lengthy. What do you think of the following?

```
EDK_HASHES = decryptMaterialsRequest.encryptedDataKeys.map(SerializeEncryptedDataKey).map(SHA512)
ENTRY_ID = SHA512(
    SHA512(cachingCMM.partitionId)
    + AlgorithmSuiteId(decryptMaterialsRequest.algorithmSuite)
    + CONCATENATE(SORTED(EDK_HASHES))
    + PADDING_OF_512_ZERO_BITS
    + SHA512(SerializeEncryptionContext(decryptMaterialsRequest.encryptionContext))
)
```

Reply:
I'm fine with either comprehensions or chaining `.map`. Thoughts @acioc ?

Reply:
Let's go with @lavaleri 's suggestion.

Reply:
Of the current implementations only JS uses a functional `map` syntax.
So while _I_ would pick I'm not sure if that is the most clear to people.

Python, Java, C all use a for loop.

https://github.com/aws/aws-encryption-sdk-c/blob/master/source/caching_cmm.c#L425
https://github.com/aws/aws-encryption-sdk-python/blob/c9619992085d3fcaddf3addb369957b74621d705/src/aws_encryption_sdk/caches/__init__.py#L99
https://github.com/aws/aws-encryption-sdk-java/blob/master/src/main/java/com/amazonaws/encryptionsdk/caching/CachingCryptoMaterialsManager.java#L364

The reason I like the functional style is that everything can be nicely inline.

```
ENTRY_ID = SHA512(
  SHA512(cachingCMM.partitionId)
  + AlgorithmSuiteId(decryptMaterialsRequest.algorithmSuite)
  + decryptMaterialsRequest
      .encryptedDataKeys
      .map(SerializeEncryptedDataKey)
      .map(SHA512)
      .sort()
      .join()
  + CONCATENATE(SORTED(EDK_HASHES))
  + PADDING_OF_512_ZERO_BITS
  + SHA512(SerializeEncryptionContext(decryptMaterialsRequest.encryptionContext))
)
```

Reply:
After a discussion offline with everybody we've decided that extracting the list comprehension into its own variable is the way to go.